text/XXX-template-poly.md

```diff
@@ -0,0 +1,154 @@
+- the indices types and constructor types do not mention the bound univs and qvar.
+
+NB: the qvar (if there is one) must be "above prop" so should not appear in relevance marks
+(it's always `Relevant`) so eg
+~~~
+Inductive prod@{q|u v|} (A:Type@{q|u}) (B:Type@{q|v}) : Type@{q|max(u,v)} :=
+  pair : A -> B -> prod A B.
+~~~
+is accepted and the arrow `A -> ...` has relevance mark `Relevant` not `RelevanceVar q`.
+See also "Note on nested templates and above Prop" below.
+
+NB': these rules mean the bound univs have `Irrelevant` cumulativity
+variance (not to be confused with relevance).
+
+## Instantiation
+
+Given a default instance (with default quality `Type`),
+a (possibly partial) application of the inductive or its constructors
+to parameters is handled by:
+
+For each BINDING parameter, we obtain the universe level (which may be algebraic)
+and quality from the actual passed parameter if available, otherwise from the default instance.
+The quality (if bound) must be "above Prop".
+
+If the quality is bound (ie non constant), it is assigned to the max of the inferred qualities
+(this is possible because of the "above Prop" assumption).
+For each bound univ level, it is assigned to the max of the corresponding obtained levels.
+
+This gives the implicit instance.
+We then check any constraints on the variables and do regular typechecking.
+
+## Subject reduction
+
+Consider `(fun X:Type@{u} => I X) (P:Type@{v}) (Q:Type@{w})`
+with `I@{i j | csts(i,j)} (p:Type@{i}) (q:Type@{j}) : Type@{f(i,j)}`
+and default instance `{i0 j0}` (which must verify `|= csts(i0, j0)`).
+
+For the beta redex to be welltyped, we must have
+- `|= csts(u, j0)`
+- `|= v <= u`
+- `|= w <= j0`
+and it has type `Type@{f(u,j0)}`.
+
+The reduced value is `I P Q`. For subject reduction to hold we need
+- `|= csts(v,w)`
+- `Type@{f(v,w)} <= Type@{f(u,j0)}`
+
+The "unbounded from below" rule means that the `csts(i,j)` are either
+- not mentioning `i` and `j` -> trivially hold from `|= csts(i0,j0)`
+  (or we extrude such constant constraints at declaration time)
+- `i <= c` with `c` constant, in which case we have `|= u <= c` from `|= csts(u,j0)` and we have `|= v <= u`
+  so `|= v <= u <= c`
+- `j <= c` with `c` constant, in which case we have `|= j0 <= c` and from `|= csts(u,j0)` and we have `|= w <= j0`
+  so `|= w <= j0 <= c`
+
+Meanwhile `f` must be built using the `max` and `+1` operators so is monotonous.
+Since `|= v <= u` and `|= w <= j0` we have `|= f(v,w) <= f(u,j0)`.
+
+The fully general proof of subject reduction should work with the same style of reasoning.
+
+## Note on nested templates and above Prop
+
+Consider
+~~~
+Inductive double@{q|u|} (A:Type@{q|u}) : Type@{q|u}
+  := Double : prod A A -> double A.
+~~~
+When checking `prod A A` we must ensure that `q` is above Prop,
+and if we had `prod A B` with A and B at different quality variables
+we would have to unify them (in elaboration) or error (in kernel).
```

Comment:
I guess `prod A B` with A and B at different qualities would land in Type thanks to above Prop

Reply:
but elaboration still need to have some unification otherwise `Inductive foo A B := wrap (_ : prod A B)` will produce `foo : _ -> _ -> Type@{Type|_}` I think